backend/ipsec/ipsec.go

```diff
@@ -0,0 +1,391 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipsec
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	log "github.com/coreos/flannel/Godeps/_workspace/src/github.com/golang/glog"
+	"github.com/coreos/flannel/Godeps/_workspace/src/github.com/vishvananda/netlink"
+	"github.com/coreos/flannel/Godeps/_workspace/src/golang.org/x/net/context"
+
+	"github.com/coreos/flannel/backend"
+	"github.com/coreos/flannel/pkg/ip"
+	"github.com/coreos/flannel/subnet"
+)
+
+const (
+	defaultReqID   = 11
+	passwordLength = 32
+	/*
+	   New IP header (Tunnel Mode)   : 20
+	   SPI (ESP Header)              : 4
+	   Sequence (ESP Header)         : 4
+	   ESP-AES IV                    : 16
+	   ESP-AES Pad                   : 0-15
+	   Pad length (ESP Trailer)      : 1
+	   Next Header (ESP Trailer)     : 1
+	   ESP-SHA-256 ICV               : 16
+	*/
+	ipsecOverhead    = 77
+	udpEncapOverhead = 8
+)
+
+type IPSECBackend struct {
+	network  string
+	password string
+	sm       subnet.Manager
+	config   *subnet.Config
+	cfg      struct {
+		Encap string
+	}
+	iked   *CharonIKEDaemon
+	lease  *subnet.Lease
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+func New(sm subnet.Manager, network string, config *subnet.Config) backend.Backend {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	i := IPSECBackend{
+		sm:      sm,
+		config:  config,
+		ctx:     ctx,
+		cancel:  cancel,
+		network: network,
+	}
+
+	return &i
+}
+
+func (i *IPSECBackend) Init(extIface *net.Interface, extIaddr net.IP, extEaddr net.IP) (*backend.SubnetDef, error) {
+	//parse configuration
+	if len(i.config.Backend) > 0 {
+		log.Info("i.config.backend length > 0")
+		if err := json.Unmarshal(i.config.Backend, &i.cfg); err != nil {
+			return nil, fmt.Errorf("error decoding IPSEC backend config: %v", err)
+		}
+	}
+
+	var err error
+
+	extLink, err := netlink.LinkByIndex(extIface.Index)
+	if err != nil {
+		return nil, fmt.Errorf("error getting default link: %v", err)
+	}
+
+	log.Info("Encap: ", i.cfg.Encap)
+
+	bytesOverhead := ipsecOverhead
+	if i.cfg.Encap != "0" {
+		bytesOverhead += udpEncapOverhead
+	}
+	log.Infof("Ipsec Overhead: %v", bytesOverhead)
+
+	newMTU := extIface.MTU - bytesOverhead
+	log.Infof("MTU set: %v", newMTU)
+
+	if err := netlink.LinkSetMTU(extLink, newMTU); err != nil {
+		return nil, fmt.Errorf("error setting MTU: %v", err)
+	}
+
+	i.iked, err = NewCharonIKEDaemon()
+	if err != nil {
+		return nil, fmt.Errorf("error creating CharonIKEDaemon struct: %v", err)
+	}
+
+	attrs := subnet.LeaseAttrs{
+		PublicIP:    ip.FromIP(extEaddr),
+		BackendType: "ipsec",
+	}
+
+	password := GenerateRandomString(passwordLength)
+	backendAttrs := subnet.BackendAttrs{
+		Password:    password,
+		BackendType: "ipsec",
+	}
+
+	log.Info("Generated password: ", password)
```

Comment:
while this is going into etcd, you probably don't want to log the actual password  -- logs get shipped all kinds of places with different ACLs, often inadequate, on them.